sodetlib/operations/bias_dets.py

```diff
@@ -201,3 +201,347 @@ def bias_to_rfrac(S, cfg, rfrac=0.5, bias_groups=None, iva=None,
+            repeat_biasstep = True
+            S.log("the biasstep script might have failed, going to repeat the measurement")
+
+    if repeat_biasstep:
+        S.log("retaking biasstep due to failure")
+        bsa_0 = bias_steps.take_bias_steps(S, cfg)
+        repeat_biasstep = False
+    
+    percentage_rn_0 = bsa_0.R0/bsa_0.R_n_IV
+    ## add in checks if detectors are SC  
+    bg_overbias_needed = []
+    overbias_needed = False
+    for bl in bias_groups:
+        mask_bg = np.where(bsa_0.bgmap==bl)[0]
+        per_bl_percentage_rn_0 = percentage_rn_0[mask_bg]
+        ## mask for SC
+        mask_sc = np.where((per_bl_percentage_rn_0 < 0.1) & (per_bl_percentage_rn_0 > -0.1))
+        ## if more than 10 percent of the detectors are SC
+        if len(mask_sc[0]) > 0.1*len(mask_bg):
+            bg_overbias_needed.append(bl)
+            overbias_needed = True
+
+    ## add in check if detectors are normal, hence we should increase the drop step
+    bg_detectors_normal = []
+    drop_from_normal = False
+    for bl in bias_groups:
+        mask_bg = np.where(bsa_0.bgmap==bl)[0]
+        per_bl_percentage_rn_0 = percentage_rn_0[mask_bg]
+        ## mask for normal
+        mask_normal = np.where((per_bl_percentage_rn_0 > 0.9) | (per_bl_percentage_rn_0 < -0.9))
+        ## if more than half of the detectors are normal
+        if len(mask_normal[0]) > 0.5*len(mask_bg):
+            bg_detectors_normal.append(bl)   
+            drop_from_normal = True
+
+
+    if overbias_needed:
+        S.log(f"some BL: {bg_overbias_needed} are stuck in SC, going to over bias them, this takes about 5mins")
+        previous_dc_biases = S.get_tes_bias_bipolar_array()
+        
+        sdl.overbias_dets(S, cfg, bias_groups=bg_overbias_needed)
+        ##force wait time 5 mins for UFM to reach equalibrium 
+        for sleep_index in range(5):
+            time.sleep(60)
+        safe_dc_biases = previous_dc_biases.copy()
+        for replace_bg in bg_overbias_needed:
+            safe_dc_biases[replace_bg] = cfg.dev.bias_groups[replace_bg]["testbed_100mK_bias_voltage"]
+            
+        S.set_tes_bias_bipolar_array(safe_dc_biases)  
+        bsa_0 = bias_steps.take_bias_steps(S, cfg)
+        bias_steps.plot_Rfrac(bsa_0)
+        
+        ## add in check if detectors are normal now
+        percentage_rn_0 = bsa_0.R0/bsa_0.R_n_IV
+        drop_from_normal = False
+        for bl in bg_overbias_needed:
+            mask_bg = np.where(bsa_0.bgmap==bl)[0]
+            per_bl_percentage_rn_0 = percentage_rn_0[mask_bg]
+            ## mask for normal
+            mask_normal = np.where((per_bl_percentage_rn_0 > 0.9) | (per_bl_percentage_rn_0 < -0.9))
+            ## if more than half of the detectors are normal
+            if len(mask_normal[0]) > 0.5*len(mask_bg):
+                bg_detectors_normal.append(bl)   
+                drop_from_normal = True
```

Comment:
This code block is repeated from before (I think). Could be tidied up to avoid repetition.

Reply:
This code block is not repeated. this code block is designed to be ran again if over_biasing is needed. After the overbiasing, the script has to choose some bias_voltage to land on, in order to be safe from dropping to SC again, the script will choose the biasvoltage recorded in the config file as "testbed_100mK_bias_voltage", this however might land the detectors in normal state. Hence this code block is for catching this after the overbiasing.

Reply:
What I'm trying to point out is that https://github.com/simonsobs/sodetlib/blob/05fea5d1b4fcdea22c099d60ce505abc5ab64626/sodetlib/operations/bias_dets.py#L325-L332 are the exact same lines as https://github.com/simonsobs/sodetlib/blob/05fea5d1b4fcdea22c099d60ce505abc5ab64626/sodetlib/operations/bias_dets.py#L289-L296

so it could be, for example, condensed into a helper function

Reply:
oh yeah yeah good point... but I am hoping it is ok to just leave it like this as it just used only twice